airsonic-main/src/main/webapp/WEB-INF/jsp/podcastChannel.jsp

```diff
@@ -1,45 +1,148 @@
-<%@ page language="java" contentType="text/html; charset=utf-8" pageEncoding="iso-8859-1"%>
+<%@ page language="java" contentType="text/html; charset=utf-8" pageEncoding="iso-8859-1" %>
 
 <%--
-  ~ This file is part of Airsonic.
-  ~
-  ~  Airsonic is free software: you can redistribute it and/or modify
-  ~  it under the terms of the GNU General Public License as published by
-  ~  the Free Software Foundation, either version 3 of the License, or
-  ~  (at your option) any later version.
-  ~
-  ~  Airsonic is distributed in the hope that it will be useful,
-  ~  but WITHOUT ANY WARRANTY; without even the implied warranty of
-  ~  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-  ~  GNU General Public License for more details.
-  ~
-  ~  You should have received a copy of the GNU General Public License
-  ~  along with Airsonic.  If not, see <http://www.gnu.org/licenses/>.
-  ~
-  ~  Copyright 2015 (C) Sindre Mehus
-  --%>
-
-<html><head>
+~ This file is part of Airsonic.
+~
+~  Airsonic is free software: you can redistribute it and/or modify
+~  it under the terms of the GNU General Public License as published by
+~  the Free Software Foundation, either version 3 of the License, or
+~  (at your option) any later version.
+~
+~  Airsonic is distributed in the hope that it will be useful,
+~  but WITHOUT ANY WARRANTY; without even the implied warranty of
+~  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+~  GNU General Public License for more details.
+~
+~  You should have received a copy of the GNU General Public License
+~  along with Airsonic.  If not, see <http://www.gnu.org/licenses/>.
+~
+~  Copyright 2015 (C) Sindre Mehus
+--%>
+
+<html>
+<head>
     <%@ include file="head.jsp" %>
     <%@ include file="jquery.jsp" %>
 
+    <script type="text/javascript" src="<c:url value='/script/utils.js'/>"></script>
     <script type="text/javascript" language="javascript">
+        var podcastEpisodes = [];
         function init() {
-            $("#dialog-delete").dialog({resizable: false, height: 170, autoOpen: false,
+            <c:forEach items="${model.episodes}" var="episode" varStatus="i">
+            var episode = {};
+            episode.id = '${episode.id}';
+            episode.mediaFileId = '${episode.mediaFileId}';
+            episode.status = '${episode.status}';
+            podcastEpisodes.push(episode);
+            </c:forEach>
```

Comment:
Not quite sure if this is the best way to pass variables from the server to client side ;-)
Is there a better way doing that (whitout using websockets)?

Reply:
This isn't websockets, this is Spring MVC (using JSP).

Here's the lifecycle of a page:
- Request comes in
- Goes to a controller
- Controller builds a _model_
- Controller forwards to a jsp template with that model (still within the server, just a different component)
- The JSP template engine picks up the JSP, and uses the model to perform server-side processing to generate an HTML
- The final HTML is sent to the client. It contains no JSP tags or even remnants of a model
- The client only interacts with this HTML. 
- If the HTML contains some calls to do things dynamically (like call other sources, including itself, such as via REST or websockets), these are now called, but this is out of the server's hand now. Now the client is in control.

The whole point of using websockets is that it allows the client to be dynamic and we don't have to go to the server again and again to generate html after every change. The client gets the HTML from the server once, and then just adapts based on the data that it retrieves from the server separately.

Reply:
I don't think this is needed

Reply:
this bit isn't needed

Reply:
Thanks for the explanation. Helps a lot! I think I mainly already understand the idea behind Spring, creating the HTML on the server side and how websocket works in general.

My question was more pointing in another direction. I was asking myself how I can send a JavaScript object (object with some attributes) from the server side (Spring, JSP) to the client side (JavaScript).
Currently the most needed information about the episode are stored in the checkbox. The index in the html-id (not needed actually) and the episode id in the html-value of the checkbox. I had to get the status (if COMPLETED) and the mediaFileId (for adding to playlist and downloading as well). Currently the getSelectedEpisodes uses jQuery to get the id of the episode aka html-value from the checkbox. I think that's what you suggested for the status as well.

To get the mediaFileId and the status I saw the following options:
a) Adding additional pseudo HTML attributes to the checkbox or somewhere in the table to have the mediaFileId and the status-value (not text) in HTML (actually I did this first, but found it very odd). But I don't wanted to bloat HTML with information actually not needed there. And I also don't wanted to "parse" the HTML text by jQuery to get the status, as this one is translated.
b) Create a new PodcastWSController and start to port the whole podcast page to use DataTables and websockets, like you did for the album view, playqueue and the playlists. By using websockets I could easily pass a JavaScript object to the client side containing all the information needed. This was to much hustle for me, for one of my firs pull requests. I thought I could do this in a second iteration.
c) What I currently did: Generate JavaScript code on the server side to represent and create the JavaScript object needed. I found this way better then creating pseudo HTML attributes, bloating the HTML or go for the whole websockets implementation. Code on the client side looks like this now when rendered:

...          
var episode = {};
episode.id = '3';
episode.mediaFileId = '';
episode.status = 'SKIPPED';
podcastEpisodes.push(episode);

var episode = {};
episode.id = '2';
episode.mediaFileId = '402';
episode.status = 'COMPLETED';
podcastEpisodes.push(episode);

var episode = {};
episode.id = '7';
episode.mediaFileId = '';
episode.status = 'SKIPPED';
podcastEpisodes.push(episode);
...

Probably I could make this a bit slimmer by just generating a JSON string on the server side instead of a whole code. And yes, the id is not needed actually, but mediaFileId and status is.

Would be interesting to have your thoughts about this. I'm not quite sure how to handle that the best way.

Reply:
Right, I'm saying you don't need this.

You don't even need additional pseudo HTML attributes, because they already exist, and we can just use them.

Scenario: you have 2 objects selected and you want to find out (a) their mediaFileIds and (b) their statuses

Pseudocode:
- getSelectedEpisodes() is the key method here
- retrieve each checkbox ("#episode"+i) and see if it's selected (we already do this)
- if it is, that's the row we need
  - episode id is just the val() of the checkbox (we already know this and use it)
  - status is a td element in the same row
    - retrieve the row: parentRow = $(checkbox).parent('tr')
    - retrieve the text of the status td box inside the parent
      - use a class as a marker or the _nth_ child selector of the parent tr
         - statusbox = parentRow.children().eq(5) or statusbox = parentRow.find('.podcast-status'), are just examples of how you could engineer it
      - compare it with the localized translation to figure out the status
          - completedStatus = '<fmt:message key="podcastreceiver.status.completed}"/>'; statusbox.text().equals(completedStatus); --> COMPLETED
  - mediafileId is given to the playButtons.jsp. One of those elements has it embedded. Again, either use _nth_ child of parent or use a class
    - starRow = parentRow.children().eq(1) or add a class and use it to locate it
    - mediaFileId is then just the suffix of the id attribute (starImage123): mid = starRow.attr('id').substring(9) -> you can make it numeric if necessary

Reply:
Ok, got it! I think I will add a class to have a better and more reliable jQuery selector to locate the stuff.

**Status**: Compare with the translated version should do the trick!
**mediaFileId**: right, it's in the playButtons, but I need to be parsed out from the HTML 
Example: "Play last" button
`<img id="add**250009**" src="icons/default_light/add.svg" alt="Play last" onclick="top.playQueue.onAdd(250009); $().toastmessage('showSuccessToast', 'Song will be played last'); return false;" style="cursor:pointer; height:18px;" title="Play last">`
==> I would take it from the id attribute as suggested

It's feasable, but I don't like "parsing" the HTML and it's attribute, as we would need to alter the selector when there's a change in the playButtons.jsp. Just sounds not very reliable to me. But anyway, I'll do it for now. When we have websockets we will have better options ;-)